### PR TITLE
mgr/volumes: sync inode attributes for cloned subvolumes

### PR DESCRIPTION
**Pull request by vshankar**

### CODE REVIEW

**Comment by vshankar** [comment]
minor fix in test

**Comment by batrick** [comment]
@vshankar ping

**Comment by vshankar** [comment]
> @vshankar ping

sorry -- was busy with other critical PRs -- will get to this today.

File: src/pybind/cephfs/cephfs.pyx [hunk 1]
```diff
@@ -1557,3 +1589,147 @@ cdef class LibCephFS(object):
...
+
+        :param path: file path for which timestamps have to be changed
+        :param times: if times is not None, it must be a tuple (atime, mtime)
+        """
+
+        self.require_state("mounted")
+        path = cstr(path, 'path')
+        if times and not isinstance(times, tuple):
+            raise TypeError('times must be a tuple')
+        actime = modtime = int(time.time())
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/cephfs/cephfs.pyx]: by default [ac]time should be the current time

File: src/pybind/cephfs/cephfs.pyx [hunk 2]
```diff
@@ -1557,3 +1589,147 @@ cdef class LibCephFS(object):
...
+        if times:
+            actime = times[0]
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/cephfs/cephfs.pyx]: right, there's not check atm -- only a check to verify if the passed in `times` is a tuple.

[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/cephfs/cephfs.pyx]: I'll update the pr

**Comment by vshankar** [comment]
@ajarr -- IIRC, were you able to run the tests locally? remember you mentioning about the tests not running?

**Comment by vshankar** [comment]
@ajarr @batrick fixed and updated -- please take a look

**Comment by vshankar** [comment]
> Looks good. Couple of questions. The new test in test_volumes passed locally using vstart_runner.
> The tests added in test_cephfs.py pass individually, but don't pass when run together. Not sure what is going on. https://pastebin.com/Q4jLb8zq

I fixed the issue in commit ce29c3c5187af883aca4759aee9dbf634bd64eee you faced when running the test locally

File: src/pybind/cephfs/cephfs.pyx [hunk 3]
```diff
@@ -1541,3 +1589,172 @@ cdef class LibCephFS(object):
...
+                               is a symbolic link (default)
+        """
+
+        self.require_state("mounted")
+        path = cstr(path, 'path')
+        if times:
+            if not isinstance(times, tuple):
+                raise TypeError('times must be a tuple')
+            if not isinstance(times[0], (int, float)):
+                raise TypeError('atime should be an float')
```
Comment (ajarr) [inline src/pybind/cephfs/cephfs.pyx]: I'm confused. Why do you check whether it's of type int or float? Checking for float should be good enough, right, as is in the error message?

File: src/pybind/cephfs/cephfs.pyx [hunk 4]
```diff
@@ -1541,3 +1589,172 @@ cdef class LibCephFS(object):
...
+            if not isinstance(times[1], (int, float)):
+                raise TypeError('mtime should be an float')
+        actime = modtime = time.time()
+        if times:
+            actime = float(times[0])
+            modtime = float(times[1])
```
Comment (ajarr) [inline src/pybind/cephfs/cephfs.pyx]: If we already check whether times[0] and times[1] are of type float, we wouldn't need this.

File: src/pybind/cephfs/cephfs.pyx [hunk 5]
```diff
@@ -1541,3 +1589,172 @@ cdef class LibCephFS(object):
...
+
+        self.require_state("mounted")
+        path = cstr(path, 'path')
+        if times:
+            if not isinstance(times, tuple):
+                raise TypeError('times must be a tuple')
+            if not isinstance(times[0], (int, float)):
+                raise TypeError('atime should be an int')
+            if not isinstance(times[1], (int, float)):
+                raise TypeError('mtime should be an int')
```
Comment (ajarr) [inline src/pybind/cephfs/cephfs.pyx]: Why do you check whether times[0] and times[1] is of type float when you say it should be of type `int` in the TypeError message?

File: src/pybind/cephfs/cephfs.pyx [hunk 6]
```diff
@@ -1541,3 +1589,172 @@ cdef class LibCephFS(object):
...
+        actime = modtime = int(time.time())
+        if times:
+            actime = int(times[0])
+            modtime = int(times[1])
```
Comment (ajarr) [inline src/pybind/cephfs/cephfs.pyx]: You wouldn't need this if you already check the types of times[0] and times[1]

**Comment by ajarr** [comment]
> > Looks good. Couple of questions. The new test in test_volumes passed locally using vstart_runner.
> > The tests added in test_cephfs.py pass individually, but don't pass when run together. Not sure what is going on. https://pastebin.com/Q4jLb8zq
> 
> I fixed the issue in commit [ce29c3c](https://github.com/ceph/ceph/commit/ce29c3c5187af883aca4759aee9dbf634bd64eee) you faced when running the test locally
> 

test_cephfs.py passed locally.
```
[rraja@bzn build]$ python3 -m nose ../src/test/pybind/test_cephfs.py
..........................
----------------------------------------------------------------------
Ran 26 tests in 23.404s
```

test_subvolume_snapshot_attr_clone also passed locally
```
2020-02-28 19:21:20,878.878 INFO:__main__:Running ['rm', '-rf', '/tmp/tmpsZtval/mnt.0']
2020-02-28 19:21:20,887.887 INFO:__main__:Running ['./bin/ceph', 'log', 'Ended test tasks.cephfs.test_volumes.TestVolumes.test_subvolume_snapshot_attr_clone']
2020-02-28 19:21:21,960.960 INFO:__main__:test_subvolume_snapshot_attr_clone (tasks.cephfs.test_volumes.TestVolumes) ... ok
2020-02-28 19:21:21,961.961 INFO:__main__:Stopped test: test_subvolume_snapshot_attr_clone (tasks.cephfs.test_volumes.TestVolumes) in 39.207761s
2020-02-28 19:21:21,961.961 INFO:__main__:
2020-02-28 19:21:21,961.961 INFO:__main__:----------------------------------------------------------------------
2020-02-28 19:21:21,961.961 INFO:__main__:Ran 1 test in 39.208s
2020-02-28 19:21:21,962.962 INFO:__main__:
2020-02-28 19:21:21,962.962 INFO:__main__:OK
```

File: src/pybind/cephfs/cephfs.pyx (hunk 5 above)
Comment (vshankar) [reply src/pybind/cephfs/cephfs.pyx]: ah! error message needs fixing.

File: src/pybind/cephfs/cephfs.pyx (hunk 6 above)
Comment (vshankar) [reply src/pybind/cephfs/cephfs.pyx]: one could pass in a floating point value but utime() has seconds granularity -- so we'd need to typecast to int.

Comment (vshankar) [reply src/pybind/cephfs/cephfs.pyx]: note that the check above is `(int, float)`

File: src/pybind/cephfs/cephfs.pyx (hunk 3 above)
Comment (vshankar) [reply src/pybind/cephfs/cephfs.pyx]: one could pass in a time value with seconds granularity too, i.e.,  tv_usec (in struct timeval) can be 0

File: src/pybind/cephfs/cephfs.pyx (hunk 4 above)
Comment (vshankar) [reply src/pybind/cephfs/cephfs.pyx]: ditto -- passed in value can be an int, but `tv_usec` needs to be filled in -- see `to_timeval()`

File: src/pybind/cephfs/cephfs.pyx (hunk 6 above)
Comment (vshankar) [reply src/pybind/cephfs/cephfs.pyx]: the other way would be to error out if float value is passed -- I think that would be the correct behaviour, but only for [f]utime() case (other call need to handle accepting int).

**Comment by vshankar** [comment]
@ajarr -- changed [f]utime() behaviour -- fails if float values are passed.

**Comment by ajarr** [comment]
http://pulpito.front.sepia.ceph.com/rraja-2020-03-02_14:19:02-fs-wip-ajarr-43965-distro-basic-smithi/

Command used
$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -k distro -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH  --filter "volumes,libcephfs_python" -N 5

**Comment by ajarr** [comment]
> http://pulpito.front.sepia.ceph.com/rraja-2020-03-02_14:19:02-fs-wip-ajarr-43965-distro-basic-smithi/
> 
> Command used
> $ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -k distro -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter "volumes,libcephfs_python" -N 5

failures are unrelated,
https://pastebin.com/VJxzZpEi
the ` test_subvolume_snapshot_attr_clone` and the python binding tests added in this PR passed

**Comment by ajarr** [comment]
> > http://pulpito.front.sepia.ceph.com/rraja-2020-03-02_14:19:02-fs-wip-ajarr-43965-distro-basic-smithi/
> > Command used
> > $ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -k distro -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter "volumes,libcephfs_python" -N 5
> 
> failures are unrelated,
> https://pastebin.com/VJxzZpEi

This wasn't entirely accurate. :-(   See, https://tracker.ceph.com/issues/44438

> the ` test_subvolume_snapshot_attr_clone` and the python binding tests added in this PR passed

**Pull request merged**